Vhost device scmi notification implement

junnan01-wu opened the pull request:
### Summary of the PR

This is Samsung Team. We want to contribute scmi notification implement.
In virtio-spec, if feature VIRTIO_SCMI_F_P2A_CHANNELS was negotiated, device should implement SCMI notification.

In our embeded environment, we enable iio sensor with notification feature, and configure this feature.

epilys commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -347,7 +347,12 @@ impl IIOSensor {
+                if self.sensor_mut().name.is_empty() {
+                    self.sensor_mut().name = name;
+                }
+            },
```
Can you give a bit more info on what this is solving? The commit message is:

> 
> It can avoid that program exit unexpectedly
> when the length of native sensor name is too long.
> 

What happens when `name` is too long and it exits?

Also, it seems like you can skip the file IO here if name is not empty:

```
    fn set_sensor_name_from_file(&mut self, path: &PathBuf) {
        // Do not override previous name
        if !self.sensor_mut().name.is_empty() {
            return;
        }
        match fs::read_to_string(path) {
            Ok(name) => {
                self.sensor_mut().name = name;
            },
```

junnan01-wu replied:
If a native iio device has it own name and too long, it will cause this error.

For example, 
We started vhost-user-scmi device like:

/usr/bin/vhost-device-scmi --socket-path /tmp/vhost-user/virtio-vm3-vscmi \
        -d iio,path=/sys/bus/iio/devices/iio:device1,channel=in_accel,name=scmi.iio.accel 

And we expected its name as "scmi.iio.accel" in guest.

But if backend can get sensor name from "/sys/bus/iio/devices/iio:device0/name", it will forcely override the name we set by command (e.g. asm330lhh_accel)

In ARM SCMI spec, the command "SENSOR_DESCRIPTION_GET" should return sensor descriptor.
And sensor_name introduction is here
	uint8 sensor_name[16]
	A NULL terminated UTF-8 format string with the sensor name, of up to 16 bytes.

Then in file "vhost-device-scmi/src/scmi.rs", the assert "v_len < size" in function "from" will fail.

We think, once backend sets sensor name in command line, it will not be overrided

epilys replied:
That makes sense. I think it might be better to check if name is not empty in `register_iio_file` when file name is `name` then, and leave `set_sensor_name_from_file` unaltered

I don't see any particular requirements in the spec about sensor names other than being 15 bytes plus the null terminator (unless it's an extended sensor name). Plus a String is allowed to have null bytes in it so it should really be a Option<CString> with a truncation...

> Then in file "vhost-device-scmi/src/scmi.rs", the assert "v_len < size" in function "from" will fail.

Maybe we should remove the assert as well for now and just ignore the extra bytes. What do you think?

junnan01-wu replied:
Your suggestion is great, maybe motificiation like this is better?

```
fn register_iio_file(&mut self, file: fs::DirEntry, axes: &mut Vec<Axis>) {
        let channel = self.channel.to_str().unwrap();
        let os_file_name = file.file_name();
        let file_name = os_file_name.to_str().unwrap_or_default();
        let raw_suffix = "_raw";
-       if file_name == "name" {
+       if file_name == "name" && self.sensor_mut().name.is_empty() {
```

And for this, you can refer to the document "[Arm System Control and Management Interface Platform Design Document](https://developer.arm.com/documentation/den0056/e?lang=en)" 
> I don't see any particular requirements in the spec about sensor names other than being 15 bytes plus the null terminator (unless it's an extended sensor name). Plus a String is allowed to have null bytes in it so it should really be a Option<CString> with a truncation...

In **_4.7.2.5.1 Sensor Descriptor_**

It define SENSOR_DESC structure

```
uint8 sensor_name[16]

A NULL terminated UTF-8 format string with the sensor name, of up to 16 bytes. 
When Bit[29] of sensor_attributes_low field is set to 1, this field contains the lower 15 bytes of the NULL terminated sensor name.
```

According to spec, if  Bit[29] of sensor_attributes_low "_Extended sensor name_" support, sensor name can be greater than 16 bytes.

And in my opinion, ignore the extra bytes instead of assert is better than current.

epilys replied:
> Your suggestion is great, maybe motificiation like this is better?

Exactly, yes!

> And in my opinion, ignore the extra bytes instead of assert is better than current.

Yep, we should use asserts when it's no longer possible to continue with a correct state, failing for this reason is not very reasonable

junnan01-wu replied:
I will re-upload this patch, and thanks for your suggestion.

junnan01-wu replied:
Hello, epilys. For this modification, I have a question about the test function
```
fn test_sensor_name_from_fs() {
        let directory = IIODirectory::new(&[("in_accel_raw", "123"), ("name", "foo")]);
        let mut sensor =
            make_iio_sensor(&directory, "in_accel".to_owned(), Some("accel".to_owned()));
        sensor.initialize().unwrap();
        assert_eq!(sensor.sensor.name, "foo");
    }
```

In previous logic, the sensor's name will be override from fs. But after my patch, it will fail.

Should this test be motidified?

junnan01-wu replied:
And the three ways to set iio sensor name should have a prority order
In my opinion, the order list should be

1.  From parameter
2.  From fs
3.  Defalut

epilys replied:
Intuitively I'd say the unit test should not overwrite the name anymore but let's ask @mz-pdm who's familiar with the crate, what do you think?

junnan01-wu replied:
That's well

mz-pdm replied:
I agree with the order list above. I can't recall any reason why the file system name should have precedence over the command line name, I probably thought only about the situation when there is no file system name.

That said, I think the three related tests (test_sensor_name_from_fs, test_sensor_name_from_params, test_default_sensor_name) should still pass. Why does test_sensor_name_from_fs fail? It should check the situation when there is a file system name but no name on the command line.

A new test should be added, similar to test_sensor_name_from_params, with `("name", "override-me")` added to IIODirectory::new. That would check that the command line name takes precedence.

junnan01-wu replied:
If with my modification 
```
-        if file_name == "name" && self.sensor.name.is_empty() {
+        if file_name == "name" && self.sensor.name == "iio" {
```
The previous unit test "_test_sensor_name_from_fs_" will failed, because it make sensor with name "accel" and 
"_assert_eq!(sensor.sensor.name, "foo")_", left is "accel“


**This is previous test, name fs exist, and make sensor with "accel" name, with my modification, it will not be override to "foo", then assert failed.**
```
fn test_sensor_name_from_fs() {
         let directory = IIODirectory::new(&[("in_accel_raw", "123"), ("name", "foo")]);
         let mut sensor =
            make_iio_sensor(&directory, "in_accel".to_owned(), Some("accel".to_owned()));
         sensor.initialize().unwrap();
         assert_eq!(sensor.sensor.name, "foo");
     }
```

mz-pdm replied:
Ah, right. Your change to test_sensor_name_from_fs is all right.

As for test_sensor_name_from_params, I would keep both the original version and your modified version (as two separate tests). This way, the tests would cover the situations when the command line name is provided and the file system name either exists (and is overridden) or doesn't exist (and the command line name still applies without any error).

junnan01-wu replied:
That's well

mz-pdm commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -62,6 +62,11 @@ const UNIT_MAPPING: &[UnitMapping] = &[
         unit_exponent: 0,
     },
+    UnitMapping {
+        channel: "in_anglvel",
```
Shouldn't this replace "in_anglevel" above? It seems "in_anglevel" doesn't exist, it's probably a typo and should be "in_anglvel" actually.

junnan01-wu replied:
Oh, you are right, I checkd variable "_iio_chan_type_name_spec_" in kernel "drivers/iio/industrialio-core.c", for iio device, it actually no "**_in_anglevel_**".
Replace previours definiation is good suggestion.
Maybe other channel name should also be checked. (For example, "**_in_angle_**" -> "**_in_angl_**")

mz-pdm commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -286,6 +296,97 @@ impl SensorT for IIOSensor {
+                        result.push(MessageValue::Signed(sensor_value_high));
+                        result.push(MessageValue::Unsigned(0));
+                        result.push(MessageValue::Unsigned(0));
+                    }
```
Shouldn't the values be output basically the same way as in reading_get? Would it be possible to utilize reading_get here?

junnan01-wu replied:
It is quite different from function _reading_get_.
In function _reading_get_, it get data from _/sys/bus/iio/devices/iio:deviceX/in_XXX_raw_

But in notificaiton flow, packaged data is from "/dev/iio:deviceX" char device.
And this device support epoll to wait data, sysfs not.

mz-pdm replied:
So the Linux IIO interfaces are quite different for explicit readings and for notifications. Readings use a separate file per each axis and notifications use a single file for all. I wonder about the other differences:
- Is there no scaling to be applied when reading the notification value (in the sense of iio:device/scale file)?
- Why do you read the value as i16 and then convert it to i64? Could the values be simply handled as i16 here and in reading_get? Or is it better to expect i64, to be future-proof?
- reading_get outputs u32, while reading_update i32. Looking into the SCMI specification, your version is correct. Should reading_get be fixed?

junnan01-wu replied:
> Is there no scaling to be applied when reading the notification value (in the sense of iio:device/scale file)?
According to scmi spec, frontend gets backend's iio axis configuration via cmd "**SENSOR_ AXIS_DESCRIPTION_GET**"
From spec **_4.7.2.6.1 Sensor Axis Descriptor_**, The bit[8] of **_uint32 axis_attributes_low_** represents that whether this sensor reports extended attributes of axis. Then extend value of "**_uint32 axis_resolution_**" will report resolution.

But function **_"fn axis_description(&self, axis: u32) -> Vec<MessageValue>"_** in currently common.rs, it always return zero. So it doesn't support getting scale now.

And what's need to note is that data from fs has same resolution with the data from /dev/iio:deviceX.
Function **reading_update** is just simply transmit data from dev without multiplied resolution. It is same with function **reading_get**.

junnan01-wu replied:
> reading_get outputs u32, while reading_update i32. Looking into the SCMI specification, your version is correct. Should reading_get be fixed?

Yes, you are right, according to spec **4.7.2.12.1 Sensor Reading Descriptor**,  the **sensor_value_low** and **sensor_value_high** should be both i32. This function should be fixed.

mz-pdm replied:
> And what's need to note is that data from fs has same resolution with the data from /dev/iio:deviceX. Function **reading_update** is just simply transmit data from dev without multiplied resolution. It is different with function **reading_get**.

This is what I don't understand: Why does `reading_get` apply scaling while `reading_update` not?

mz-pdm replied:
Is it perhaps because the scan_type format specification contains the scaling (shift), which is 0 in this case?

mz-pdm replied:
> I don't know whether i64 is better, and why read_axis return i64.

I think I used i64 because this is what the SCMI specification supports. I think it makes sense in reading_get. As for reading_update, it could be simplified a bit assuming the supported value is i16 or left as it is. Up to you, depending what you decide to do about the supported formats.

mz-pdm replied:
> according to spec **4.7.2.12.1 Sensor Reading Descriptor**, the **sensor_value_low** and **sensor_value_high** should be both i32. This function should be fixed.

Would you prefer to fix it in your patches or should I post the fix separately?

junnan01-wu replied:
> This is what I don't understand: Why does reading_get apply scaling while reading_update not?

funcation _reading_get_ is not implemented by us, in this function, it calls _read_axis_ to get data.
```
fn read_axis(&self, axis: &Axis) -> Result<i64, ScmiDeviceError> {
        let path_result = axis.path.clone().into_string();
        let mut value: i64 =
            read_number_from_file(Path::new(&(path_result.unwrap() + "_raw")))?.unwrap();
        let offset: Option<i64> = self.read_axis_offset(&axis.path)?;
        if let Some(offset_value) = offset {
            match value.checked_add(offset_value) {
                Some(new_value) => value = new_value,
                None => {
                    error!(
                        "IIO offset overflow in {:?}: {} + {}",
                        &axis.path,
                        value,
                        offset.unwrap()
                    );
                    return Err(ScmiDeviceError::GenericError);
                }
            }
        }
        let scale: Option<f64> = self.read_axis_scale(&axis.path)?;
        if let Some(scale_value) = scale {
            let exponent_scale = 10.0_f64.powi(i32::from(axis.custom_exponent));
            value = (value as f64 * (scale_value / exponent_scale)).round() as i64;
        }
        Ok(value)
    }
```
But I don't think the operation multiplied with scale_value should be done here. It should just return raw data from fs.
And the frontend's scale value should return by command **SENSOR_ DESCRIPTION_GET** or **SENSOR_ AXIS_DESCRIPTION_GET**

Then frontend can get real value by multiplied rawdata with scale 

It's my opinion.

junnan01-wu replied:
> Would you prefer to fix it in your patches or should I post the fix separately?
Both ok, I can update it with a single patch and rebase it in this pull request.

mz-pdm replied:
> I can update it with a single patch and rebase it in this pull request.

OK, thank you.

mz-pdm replied:
> But I don't think the operation multiplied with scale_value should be done here. It should just return raw data from fs.
> And the frontend's scale value should return by command **SENSOR_ DESCRIPTION_GET** or **SENSOR_ AXIS_DESCRIPTION_GET**

I think this is what `custom_exponent` function ensures. But IIRC there was no assurance that the IIO scale provided by the fs is a power of 10. Hence the additional scaling in reading_get/read_axis. Maybe it's only a theoretical problem, I don't know, but I wanted to be on the safe side. Whatever we decide to do, I'd like to have reading_get and reading_update doing the same thing, having divergent computations would be confusing and maybe incorrect in one of them.

junnan01-wu replied:
I find a problem that, in our embedded environment, we use ASM330LHH sensor, and it does not provide
suhc "**in_anglvel_X_scale**" for user, it only provide "**in_anglvel_scale**".
It will cause that function **_"read_axis_scale"_** always return _**None**_. It finally causes that **_scale_** is _**None**_

Does function "**_reading_get_**" should consider this?

junnan01-wu replied:
I make a modification like this to deal this situation.

```
diff --git a/vhost-device-scmi/src/devices/iio.rs b/vhost-device-scmi/src/devices/iio.rs
index 25e66ea..9066307 100644
--- a/vhost-device-scmi/src/devices/iio.rs
+++ b/vhost-device-scmi/src/devices/iio.rs
@@ -522,6 +522,7 @@ impl IIOSensor {
    fn read_axis_file<T: FromStr>(
        &self,
        path: &OsStr,
        name: &str,
    ) -> Result<Option<T>, ScmiDeviceError> {
         for value_path in [
             Path::new(&(String::from(path.to_str().unwrap()) + "_" + name)),
             &Path::new(&path).parent().unwrap().join(name),
+            Path::new(&format!("{}_{}",&String::from(path.to_str().unwrap())[..path.len()-2],name)),
         ]
         .iter()
         {
```
So that it can find all kinds of "scale" attribute

- /sys/bus/iio/devices/iio:device0/in_anglvel_y_scale
- /sys/bus/iio/devices/iio:device0/scale
- /sys/bus/iio/devices/iio:device0/in_anglvel_scale

Previous, it can only deal the first two cases.

junnan01-wu replied:
> I think this is what custom_exponent function ensures. But IIRC there was no assurance that the IIO scale provided by the fs is a power of 10. Hence the additional scaling in reading_get/read_axis. Maybe it's only a theoretical problem, I don't know, but I wanted to be on the safe side. Whatever we decide to do, I'd like to have reading_get and reading_update doing the same thing, having divergent computations would be confusing and maybe incorrect in one of them.

I analysis the scmi kernel and iio drivers, and your work is right.
Because the in _**axis_attributes_high**_, only has Exponent this single value. It means that in frontend scmi iio sysfs "in_XXX_scale" is must a value power-of-10 multiplier.
Therefore, no matter _reading_get_ or _reading_update_, it must return agent a value multiplied with platform's scale.

> Bits[15:11] Exponent: The power-of-10 multiplier in two’s-complement format that is applied to the sensor unit specified by the SensorType field.

For example.
In backend, rustvmm reads "**in_anglvel_x_raw**" and return "53", and reads "**in_anglvel_scale**" and return "0.000152716"
then function "**_custom_exponent_**" will returns "-4" here. The real physical data should be "53*0.000152716=0.008093948"

And frontend reads "**in_anglvel_x_scale**" will return "0.000100000" (because custom_exponent is -4). therefore the raw data from "**in_anglvel_x_raw**" should return "80" (So that 80*0.000100000 is closer to the real physical value)

Am I analysis it correct? I will be very honored to refer your opinion.

If yes, I will refactor "_reading_update_" to do same things with "_reading_get_"

mz-pdm replied:
Yes, I think we understand each other now in this matter.
As for sysfs scale files, does it matter the order in which the files are looked up or is only one of the in_anglevel_y_scale, in_anglevel_scale, and scale present?

junnan01-wu replied:
I am not sure whether order is important, but in theory, it will not exist more than one "scale" sysfs for one axis.
I will research more in kernel, maybe answer can be found out.

mz-pdm commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -286,6 +296,97 @@ impl SensorT for IIOSensor {
+        //future change to self.get_data_format
+        //cat /sys/bus/iio/devices/iio:deviceX/scan_elements/in_XXX_type.
+        //for example le:s16/16>>0 , [x,y,z,(t)] each is signed 16, and Little Endian
+        //so one sample length is (3 or 4)*(16/8) byte
```
I'm not sure I understand it correctly, let me rephrase it and please correct it if it is wrong:

```
// The buffer length should correspond to the IIO device type.
// The type is available from /sys/bus/iio/devices/iio:deviceX/scan_elements/in_XXX_type.
// For example, if the content of in_XXX_type is le:s16/16>>0, each value is a little endian
// signed 16-bit integer. For a 3-axes sensor with [x, y, z, (t)] values, i.e. the 3 axes plus an
// optional timestamp, we need 6 or 8 bytes buffer.
```

junnan01-wu replied:
Perfect rephrase, exactly yes!

junnan01-wu replied:
Because rust-vmm backend is not clear about the package format of data from /dev/iio:deviceX.
So it's little difficult to re-package it and satisfy SCMI spec requirements.
Therefore, in our implement, we only simply delt like this and ignore timestamp.

In order to consider timestamp. The SENSOR_CONFIG_SET command should also be refactored.
According to scmi spec, the parameter "sensor_config" in command SENSOR_CONFIG_SET should stores more information. Maybe we should also add a variable into SensorT to store configuration of each sensor.

> Bit[1] Timestamp reporting:
> Set to 1 if the sensor value provided by the platform should be timestamped.
> Set to 0 if the sensor value provided by the platform should not be timestamped.
> If the sensor does not support timestamp reporting or its configuration, this bit should be set to 0.

mz-pdm replied:
If timestamps are to be used then yes. Otherwise it's optional, since the current bit value 0 is right for non-timestamped values. Up to you.

mz-pdm commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -286,6 +296,97 @@ impl SensorT for IIOSensor {
+        //for example le:s16/16>>0 , [x,y,z,(t)] each is signed 16, and Little Endian
+        //so one sample length is (3 or 4)*(16/8) byte
+        let sample_buffer_len = 8;
+        let mut buffer = vec![0u8; sample_buffer_len];
```
This feels quite unsafe. How about checking that the value type is indeed le:s16 and returning an error in case it is not, rather than outputting nonsense in such a case?

junnan01-wu replied:
Is it better to read from "in_XXX_type" firstly? And re-define related arguments according to result?
I will try to refactor it.

mz-pdm replied:
Yes, that would be best.

mz-pdm commented on vhost-device-scmi/src/main.rs:
```diff
@@ -126,6 +126,11 @@ fn start_backend(config: VuScmiConfig) -> Result<()> {
         )
         .unwrap();
 
+        // Register devices such as "/dev/iio:deviceX" which can initiatively notify frontend to epoll handler.
```
"..., which can actively notify the frontend to epoll the handler." ?

junnan01-wu replied:
well

mz-pdm commented on vhost-device-scmi/src/main.rs:
```diff
@@ -126,6 +126,11 @@ fn start_backend(config: VuScmiConfig) -> Result<()> {
         .unwrap();
 
+        // Register devices such as "/dev/iio:deviceX" which can initiatively notify frontend to epoll handler.
+        // Then once there are data coming from these devices, it will occur event. (device_event=3)
```
"Then once there is data coming from these devices, an event will be created. (device_event=3)"

junnan01-wu replied:
Well

mz-pdm commented on vhost-device-scmi/src/scmi.rs:
```diff
@@ -56,8 +58,9 @@ pub type MessageValues = Vec<MessageValue>;
-    Command,     // 0
-    Unsupported, // anything else
+    Command,          // 0
+    Notification = 3, //Notifications have a message type of 3
```
Please add a space after //.

junnan01-wu replied:
Done

mz-pdm commented on vhost-device-scmi/src/scmi.rs:
```diff
@@ -126,6 +129,15 @@ impl From<&MessageValues> for Response {
 }
 
+impl Response {
+    // Notification is different with general response, it doesn't need ReturnStatus.
```
s/with/from/ ?

junnan01-wu replied:
done

mz-pdm commented on vhost-device-scmi/src/scmi.rs:
```diff
@@ -247,6 +259,7 @@ pub const SENSOR_AXIS_DESCRIPTION_GET: MessageId = 0x7;
 pub const SENSOR_CONFIG_GET: MessageId = 0x9;
 pub const SENSOR_CONFIG_SET: MessageId = 0xA;
 pub const SENSOR_CONTINUOUS_UPDATE_NOTIFY: MessageId = 0xB;
+pub const SENSOR_UPDATE: MessageId = 0x1;
```
I would add the following comment above this line to avoid confusion about the message id numbers:
``// Notification message ids:``

junnan01-wu replied:
I add it in next push

mz-pdm commented on vhost-device-scmi/src/scmi.rs:
```diff
@@ -575,6 +590,26 @@ pub trait ScmiDevice: Send {
+
+    /// Get notify message from this device.
+    ///
+    /// Usually need to refine it for different protocal.
```
s/refine/redefine/; s/protocal/protocol/

junnan01-wu replied:
Well

mz-pdm commented on vhost-device-scmi/src/scmi.rs:
```diff
@@ -575,6 +590,26 @@ pub trait ScmiDevice: Send {
+    ///
+    /// Usually need to refine it for different protocal.
+    /// device_index is the index of this device in its protocol list.
+    /// message_id points that which notification it should be sent.
```
"message_id is the notification message id to be used." ?

junnan01-wu replied:
Oh, yes. For notification in sensor protocol, spec defines two commands.

> SENSOR_TRIP_POINT_EVENT -> message_id=0
> SENSOR_UPDATE -> message_id = 1

Currently, we only implement for "SENSOR_UPDATE".

mz-pdm commented on vhost-device-scmi/src/scmi.rs:
```diff
@@ -666,11 +731,71 @@ impl ScmiHandler {
+    /// According to device_event, find out the device which will do notification.
+    /// Then call its notify function to return a ScmiResponse.
+    /// Now only SENSOR PROTOCOL can do notification.
+    /// And it only support SENSOR_UPDATE meesage.
```
"And it supports only SENSOR_UPDATE message."

epilys replied:
Plus it's good formatting style to use backticks for uppercase constant names:

```markdown
And it supports only the `SENSOR_UPDATE` message.
```

junnan01-wu replied:
All right

mz-pdm commented on vhost-device-scmi/src/vhu_scmi.rs:
```diff
@@ -57,6 +60,8 @@ pub enum VuScmiError {
     InsufficientDescriptorSize(usize, usize),
     #[error("Failed to send notification")]
     SendNotificationFailed,
+    #[error("Failed to get nofiticaiton data from scmi device")]
```
s/notificaiton/notification/

junnan01-wu replied:
Well

mz-pdm commented on vhost-device-scmi/src/vhu_scmi.rs:
```diff
@@ -132,6 +137,19 @@ impl VuScmiBackend {
         })
     }
 
+    /// This function is used to register all devices which has can be notified into handler.
```
"/// Registers all the devices that can provide notifications." ?

junnan01-wu replied:
Well

mz-pdm reviewed (commented):
Thank you for the patches. An initial high-level review in comments. Most notably, I leave the polling and notification parts for review to those who understand what's the right to do there.

junnan01-wu commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -230,6 +232,15 @@ pub struct Sensor {
     /// using [Sensor::new] are disabled initially.
     enabled: bool,
+
+    /// Sensor notification can be enabled or disabled when frontend send SENSOR_CONTINUOUS_UPDATE_NOTIFY.
```
All right

junnan01-wu commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -230,6 +232,15 @@ pub struct Sensor {
+
+    /// Sensor notification can be enabled or disabled when frontend send SENSOR_CONTINUOUS_UPDATE_NOTIFY.
+    notify_enabled: bool,
+    /// If this sensor support notifing fronted initiatively, it should record
```
All right

junnan01-wu commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -230,6 +232,15 @@ pub struct Sensor {
+    /// File there.
+    pub notify_dev: Option<File>,
+
+    /// Store this sensor's id in Strcut Sensor
```
In file "vhost-device-scmi/src/devices/scmi.rs", we refactor function "insert" to return a identify.
The structure "DeviceIdentify" contains "protocol id, device id", it is the key in hashmap for us to find out which device is notifying. (By protocol id, and device id, it can locate specified device).

mz-pdm replied:
I see. How about to change the doc string to let's say:
``/// Sensor id, to identify the sensor in notification lookup.``

junnan01-wu replied:
I will change it.

junnan01-wu commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -442,7 +465,13 @@ pub trait SensorT: Send {
+                if parameters[0].get_unsigned() == 1 {
+                    self.sensor_mut().notify_enabled = true;
+                    Ok(vec![MessageValue::Signed(1)])
+                } else {
```
In SCMI spec, it defines the parameters of cmd "SENSOR_CONTINUOUS_UPDATE_NOTIFY" in 4.7.2.13 "[Arm System Control and Management Interface Platform Design Document](https://developer.arm.com/documentation/den0056/e?lang=en)"

The requirement of paramter "uint32 notify_enable" is like following
```
Bits[31:1] Reserved, must be zero.
Bit[0] Notify enable:
If this value is 0, the platform does not send any SENSOR_UPDATE notifications to the agent.
If this value is set to 1, the platform sends SENSOR_UPDATE notifications to the agent.
See section 4.7.4.2 for more details about the SENSOR_UPDATE notification.
```

 At the same time, in linux kernel, function "_scmi_sensor_request_notify_" in "_arm_scmi/sensors.c_"  also satify this requirement. 
```
static inline int
scmi_sensor_request_notify(const struct scmi_protocol_handle *ph, u32 sensor_id,
                           u8 message_id, bool enable)
{
        int ret;
        u32 evt_cntl = enable ? SENSOR_NOTIFY_ALL : 0;
```
And we can make a conclusion that, this command paramter has no other invaild value.

mz-pdm replied:
Not all clients must work properly, there can be software bugs or hardware errors. I'd recommend to return an error if the value is not 0 or 1.

junnan01-wu replied:
I refactored this flow to handle unexpected situation you said

mz-pdm replied:
Thank you, looks good now.

junnan01-wu commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -442,7 +465,13 @@ pub trait SensorT: Send {
             SENSOR_CONTINUOUS_UPDATE_NOTIFY => {
                 // Linux VIRTIO SCMI insists on this.
                 // We can accept it and ignore it, the sensor will be still working.
```
Yes, this comment should be modified.
Maybe you can leave the suggestion comment here, and I will add it.

mz-pdm replied:
I would change the comment to just a single line:
``// Linux VIRTIO SCMI insists on handling this.``

junnan01-wu replied:
well

mz-pdm commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -396,7 +396,9 @@ impl IIOSensor {
-        if file_name == "name" {
+        // Only override name when no defined by parameter.
+        // If both fs and parameter exist, use parameter first and skip from file.
+        if file_name == "name" && self.sensor.name == "iio" {
```
Hm, if somebody gets an idea to use "iio" as the name on the command line, this won't work and will cause confusion. Changing the default name to something less likely to be used (something like "_iio_default" or "(IIO)") would be the easiest fix. Another option is to track whether the default name was used but it wouldn't be very elegant.

junnan01-wu replied:
Because there is a default name "iio", then if someboay want to use "iio" as the name on command line. It will still be "iio".
But your idea is good, it actually confuse developer to read this code here.
Should we make a macro to define default name?

For example, at the beginning of the file, define a macro like below
```
diff --git a/vhost-device-scmi/src/devices/iio.rs b/vhost-device-scmi/src/devices/iio.rs
index 97d23a4..43b88f4 100644
--- a/vhost-device-scmi/src/devices/iio.rs
+++ b/vhost-device-scmi/src/devices/iio.rs
@@ -163,6 +163,8 @@ const UNIT_MAPPING: &[UnitMapping] = &[
     },
 ];

+const IIO_DEFAULT_NAME: &str = "iio";
+
 /// Representation of an IIO channel axis.
 ///
 /// Used also for scalar values.
@@ -433,7 +435,7 @@ impl IIOSensor {
     #[allow(clippy::new_ret_no_self)]
     pub fn new(properties: &DeviceProperties) -> Result<Self, DeviceError> {
         properties.check(&["path", "channel"], &["name"])?;
-        let sensor = Sensor::new(properties, "iio");
+        let sensor = Sensor::new(properties, IIO_DEFAULT_NAME);
         Ok(Self {
             sensor,
             path: OsString::from(properties.get("path").unwrap()),
@@ -502,7 +504,7 @@ impl IIOSensor {
         let raw_suffix = "_raw";
         // Only override name when no defined by parameter.
         // If both fs and parameter exist, use parameter first and skip from file.
-        if file_name == "name" && self.sensor.name == "iio" {
+        if file_name == "name" && self.sensor.name == IIO_DEFAULT_NAME {
             self.set_sensor_name_from_file(&file.path());
         } else if file_name.starts_with(channel) && file_name.ends_with(raw_suffix) {
```

It seems better to understand this logic, do you think so?

mz-pdm replied:
> Because there is a default name "iio", then if someboay want to use "iio" as the name on command line. It will still be "iio". 

It won't because the name will be read from the file system in such a case.

> It seems better to understand this logic, do you think so?

Yes, it is in any case.

junnan01-wu replied:
All right.

junnan01-wu commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -453,6 +482,68 @@ pub trait SensorT: Send {
         }
     }
+
+    /// Returns the notification messages from device.
```
well

junnan01-wu commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -453,6 +482,68 @@ pub trait SensorT: Send {
+
+    /// Returns the notification messages from device.
+    ///
+    /// Usually need to redefine this. Different sensor may has different way to
```
Maybe this is better

> s/Different sensor may has different way/Different sensors may have different ways/

mz-pdm replied:
Yes, this is what I meant.

junnan01-wu commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -453,6 +482,68 @@ pub trait SensorT: Send {
+    /// Returns the notification messages from device.
+    ///
+    /// Usually need to redefine this. Different sensor may has different way to
+    /// get notification.
```
well

junnan01-wu commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -453,6 +482,68 @@ pub trait SensorT: Send {
+    ///
+    /// Usually need to redefine this. Different sensor may has different way to
+    /// get notification.
+    fn reading_device(&self, _device_index: u32) -> DeviceResult {
```
It's a good suggestion. This function is used to read update data from native device.
_reading_update_ seems better

junnan01-wu commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -453,6 +482,68 @@ pub trait SensorT: Send {
+    /// Enable Sensor notify function.
+    ///
+    /// Usually need to redefine this. Different sensor will do different
+    /// configuration set to enable notify funciton.
```
Well

junnan01-wu commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -453,6 +482,68 @@ pub trait SensorT: Send {
+    /// Disable Sensor notify function.
+    ///
+    /// Usually need to redefine this. Different sensor will do different
+    /// configuration set to disable notify funciton.
```
well

junnan01-wu commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -453,6 +482,68 @@ pub trait SensorT: Send {
+
+    /// Get notify device fd
+    ///
+    /// This fd is used for getting notification.
```
well

junnan01-wu commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -453,6 +482,68 @@ pub trait SensorT: Send {
+        None
+    }
+
+    /// Set a id for this device.
```
well

junnan01-wu commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -453,6 +482,68 @@ pub trait SensorT: Send {
+
+    /// Get a nofication message value from this sensor.
+    ///
+    /// The sensor's default implement only support SENSOR_UPDATE notification.
```
Oh sorry, in currently, we only implement this notification.
And there still remains "SENSOR_TRIP_POINT_EVENT" notification which we haven't implemented.
We are not familiar with this command, maybe it can be implemented in future.

junnan01-wu replied:
well

junnan01-wu commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -237,6 +241,12 @@ impl SensorT for IIOSensor {
         axes.sort_by(|a1, a2| a1.path.cmp(&a2.path));
         self.axes = axes;
+
+        // Open /dev/iio:deviceXX and store in SensorT
```
Well

junnan01-wu commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -237,6 +241,12 @@ impl SensorT for IIOSensor {
+        // Open /dev/iio:deviceXX and store in SensorT
+        let path_split: Vec<_> = self.path.to_str().unwrap().split('/').collect();
+        let dev_path = "/dev/".to_owned() + path_split[path_split.len() - 1];
+        self.sensor_mut().notify_dev = Some(File::open(dev_path.clone()).unwrap());
```
Yes, according to the iio industrial architecture (You can refer to [this](https://www.kernel.org/doc/html/latest/driver-api/iio/core.html)) 

> There are two ways for a user space application to interact with an IIO driver.
> **/sys/bus/iio/iio:deviceX/**, this represents a hardware sensor and groups together the data channels of the same chip.
> **/dev/iio:deviceX**, character device node interface used for buffered data transfer and for events information retrieval.

If sensor implement notification, all sample data and events are comming for "/dev/iio:deviceX"

mz-pdm replied:
OK. I can see you already wrapped it by a check for the file presence, good.

junnan01-wu replied:
Yes, I open it after check it really exist

junnan01-wu commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -286,6 +296,97 @@ impl SensorT for IIOSensor {
+                .map_err(|e| DeviceError::IOError(scan_path.into(), e))?;
+        }
+        Ok(())
+    }
```
Maybe you don't notice the details.
The enable flow is different from disable flow.
When frontend want to enable it, the order is
```
echo 1 > /sys/bus/iio/devices/{}/scan_elements/{}_x_en
echo 1 > /sys/bus/iio/devices/{}/scan_elements/{}_y_en
echo 1 > /sys/bus/iio/devices/{}/scan_elements/{}_z_en
echo 1 > /sys/bus/iio/devices/{}/buffer/enable
```
But disable flow is converse.
 ```
echo 0 > /sys/bus/iio/devices/{}/buffer/enable
echo 0 > /sys/bus/iio/devices/{}/scan_elements/{}_x_en
echo 0 > /sys/bus/iio/devices/{}/scan_elements/{}_y_en
echo 0 > /sys/bus/iio/devices/{}/scan_elements/{}_z_en
```

junnan01-wu replied:
But refactor this to a comman function is a good suggestion, I will try to do it

mz-pdm replied:
Ah, right, there is the reverse order of the operations. But even then the code is almost identical, so there should be a common function.

junnan01-wu replied:
I refatored a new comman function "notify_status_set"
in branch from [00acf23 ](https://github.com/rust-vmm/vhost-device/commit/00acf23a028ef5f4ac6dd2095f093263019066b6)to [fd4d28e](https://github.com/rust-vmm/vhost-device/commit/fd4d28ea02f5a9eb0476cd93ffb36a2e47d36b30)

Is this better than before?

mz-pdm replied:
Yes, thank you.

junnan01-wu commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -286,6 +296,97 @@ impl SensorT for IIOSensor {
+                        buffer[self.axes.len() * 2..self.axes.len() * 2 + 2]
+                            .try_into()
+                            .unwrap(),
+                    );
```
Oh, it's a testing code I remain there.
In _scan_elements_, there is a attribute "_in_timestamp_en_". In theory, if we enable it, data comming for device should appending with timestamp.
But I in test it in our environment, there is no timestamp data, it really confused me. Therefore I remain there for future use.
Maybe it should be tested by other iio sensor.

mz-pdm replied:
I see.  In such I case, I would suggest to replace the given piece of code with a comment explaining the situation.

junnan01-wu replied:
Ok, you are right, I will comment this code.

junnan01-wu commented:
Update file "coverage_config_x86_64.json" to pass vhost-device-ci

mz-pdm commented:
> Update file "coverage_config_x86_64.json" to pass vhost-device-ci

Oh, reducing coverage_score is not very good. New code should be covered by unit tests if reasonably possible.

junnan01-wu commented:
It's actually not a good idea, I will add new unit tests for new code in next push for order not reducing coverage rate

junnan01-wu commented:
In [force-pushed](https://github.com/rust-vmm/vhost-device/compare/fd4d28ea02f5a9eb0476cd93ffb36a2e47d36b30..bd6f00e684e6a5e9e23aac5641af1a4ec202bb14) the vhost-device-scmi branch from [fd4d28e ](https://github.com/rust-vmm/vhost-device/commit/fd4d28ea02f5a9eb0476cd93ffb36a2e47d36b30)to bd6f00e, I did following works

1. Modify "**SENSOR_READING_GET**" flow in handle.
After supporting notification, once frontend wants to read from sysfs "_XXX_raw_", it should call "_notify_status_set_" to disable notify before "reading_get", otherwise, it will return "device busy".

2. Add struct "**_ChanScanType_**" in struct "**_Axis_**"
In notification flow, data from /dev/iio:deviceX is formatted according to the rule from "_scan_elements/XXX_type_". Therefore, each axis should store the parsing result from it.
And I add a "new" implmentment for Axis, if scan_type exists, it will parse it and stores in self.

3. As we communicated previously, I modified the return type of function "_reading_get_".
According to SCMI spec, Sensor Reading Descriptor is (i32,i32,u32,u32).
Correspondingly, unit test of "_reading_get_" also modified.

4. Refactor function "_read_axis_".
Previously, function "_read_axis_" has two mainly processes.
The first is reading data from "_XXX_raw_"
The second is reading offset and scale, then multiplying raw with them
The second flow can be reused by "_reading_update_", so I sperate it to a new function "_deal_axis_raw_data_", so that "_reading_update_" can call it. 

5. In "_reading_update_", initialize buffer size according to scan_type.

6. I squash last two commits, they are both for notification flow.

7. Add som unit tests.

mz-pdm commented:
Thank you for the update, it looks nice at the first glance. I'll try to review the whole PR again in the coming days. And sorry for a delayed response, I was away.

junnan01-wu commented:
In push branch from [f6de7cf ](https://github.com/rust-vmm/vhost-device/commit/f6de7cf71944d1ac7a313c727f8a64bae062b071)to [c80b59f ](https://github.com/rust-vmm/vhost-device/commit/c80b59fb7634c311e5313bf05e59fc6f23de512b)
 

I change Sensor::name to Option.
Related test function also changed.

junnan01-wu commented:
In push branch from from [c80b59f ](https://github.com/rust-vmm/vhost-device/commit/c80b59fb7634c311e5313bf05e59fc6f23de512b)to [02f5204](https://github.com/rust-vmm/vhost-device/commit/02f520470dd5907ec2ecdb44894396a116009cd9)

Some units test for notification flow have added.

Then, coverity score will not decrease.

mz-pdm commented:
Thank you for all your changes. I'll be unresponsive the next 1-2 weeks but don't worry, I won't forget about your PR.
Anyway, a good progress has been made and I think it's time to involve other reviewers or the maintainers now.

junnan01-wu commented:
Thanks for your kindly review. It really help us a lot.

stefano-garzarella commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -224,19 +224,18 @@ pub fn devices_help() -> String {
     /// The sensor name (possibly truncated) as reported to the guest.
-    pub name: String,
     /// Whether the sensor is enabled.
+    pub name: Option<String>,
```
Why we are moving this under the comment of the next option?

junnan01-wu replied:
For each scmi device, it should have name, but there has several ways to get name.

> 1. Use name parameter from user's command.
> 2. Use sysfs or other native attributes.
> 3. Use default name

At the beginning of creating device, if no user's command with name, we don't know whether default name should be set, because we have not scanned the sysfs to find whether there exist name attritbutes yet.
So, after discussing with [mz-pdm](https://github.com/mz-pdm), we think it's better to change this var to Option, so that it can help us to judge whether name has been set when register.

stefano-garzarella replied:
I'm just asking why moving that line in the comment on the next field of the structure.
```
@@ -224,8 +224,8 @@ pub fn devices_help() -> String {
 #[derive(Debug)]
 pub struct Sensor {
     /// The sensor name (possibly truncated) as reported to the guest.
-    pub name: String,
     /// Whether the sensor is enabled.
+    pub name: Option<String>,
     ///
     /// Sensors can be enabled and disabled using SCMI.  [Sensor]s created
     /// using [Sensor::new] are disabled initially.
```

junnan01-wu replied:
Oh, sorry. It is a mistake, I will modify it.

junnan01-wu replied:
In branch from [1fb2f52 ](https://github.com/rust-vmm/vhost-device/commit/1fb2f52e500f9fe6821e31a54fdf1f8c231c9b72)to [533416e](https://github.com/rust-vmm/vhost-device/commit/533416e21d45c7cc4b82072e6a146ac9bb82456f), I fix this mistake.

stefano-garzarella commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -323,7 +322,7 @@ pub trait SensorT: Send {
             self.format_unit(0)
         };
-        let name = self.sensor().name.clone();
+        let name = self.sensor().name.clone().unwrap();
```
can the `name` be `None` and panic here?

junnan01-wu replied:
Yes, it will panic here. If there is no 1, and 2 which I [mentioned](https://github.com/rust-vmm/vhost-device/pull/710#discussion_r1747849968).
We will use default name to promise that it is not **None**.

stefano-garzarella replied:
okay, so maybe a comment here would be nice.

junnan01-wu replied:
In branch from [1fb2f52 ](https://github.com/rust-vmm/vhost-device/commit/1fb2f52e500f9fe6821e31a54fdf1f8c231c9b72)to [533416e](https://github.com/rust-vmm/vhost-device/commit/533416e21d45c7cc4b82072e6a146ac9bb82456f),
I leave a comment here.

stefano-garzarella commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -159,6 +159,8 @@ const UNIT_MAPPING: &[UnitMapping] = &[
     },
 ];
 
+const IIO_DEFAULT_NAME: &str = "iio";
```
Should we document this default in the help page and README?

junnan01-wu replied:
Ok, I will document it in help page and README in next push

junnan01-wu replied:
If a user of vhost-device-scmi sets name with command parameter, it will use his setting. If not, it will use name from sysfs name attribute, otherwise, it will use this default name.

In my opinion, I think we can skip this, it will not influcene user. It is a guarantee for scmidevice to have a name.

stefano-garzarella commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -280,13 +398,110 @@ impl SensorT for IIOSensor {
             let value = self.read_axis(axis)?;
-            result.push(MessageValue::Unsigned((value & 0xFFFFFFFF) as u32));
-            result.push(MessageValue::Unsigned((value >> 32) as u32));
+            result.push(MessageValue::Signed((value & 0xFFFFFFFF) as i32));
+            result.push(MessageValue::Signed((value >> 32) as i32));
```
Are these changes related?
Or it's a fix of a pre-existing issue?

junnan01-wu replied:
Yes, it is a fix of a pre-existing issue which I have discussed with [mz-pdm](https://github.com/mz-pdm).
For more details, you can look this [conversation](https://github.com/rust-vmm/vhost-device/pull/710#discussion_r1722857750)

stefano-garzarella replied:
So, please put in a separate commit explaining what you're fixing. (same PR is fine)

junnan01-wu replied:
In latest push, I have separated it.

stefano-garzarella commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -390,11 +605,11 @@ impl IIOSensor {
-        axes.push(Axis {
-            path: OsString::from(path),
+        axes.push(Axis::new(
+            OsString::from(path),
             unit_exponent,
             custom_exponent,
-        });
+        ));
```
Ditto, are this change related?

junnan01-wu replied:
No, this change is a new feature.
I implement the function "new" for "Axis", it can help us to get "scan_type" from sysfs for each axis, if this attribute exists. It is used for notification flow.

junnan01-wu replied:
No, this change is a new feature.
I implement the function "new" for "Axis", it can help us to get "scan_type" from sysfs for each axis, if this attribute exists.

stefano-garzarella commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -453,8 +673,12 @@ impl IIOSensor {
+        self.deal_axis_raw_data(value, axis)
+    }
+
+    fn deal_axis_raw_data(&self, mut value: i64, axis: &Axis) -> Result<i64, ScmiDeviceError> {
```
Can you do this kind of refactoring in a separate commit?
This patch is very big, so please try to split at least the refactoring in other patches.

junnan01-wu replied:
Oh...Yes, this patch is very big. I will try to separate it in next push.

junnan01-wu replied:
I have separated it into two patches

stefano-garzarella commented on vhost-device-scmi/src/vhu_scmi.rs:
```diff
@@ -326,6 +347,68 @@ impl VuScmiBackend {
+                .map_err(|_| VuScmiError::SendNotificationFailed);
+        } else {
+            debug!("Get Notification Failed!");
+            return Ok(());
```
we can remove this return (if we really want to return `Ok(())`)

junnan01-wu replied:
You are right, I will remove it.

stefano-garzarella reviewed (commented):
I don't know much about scmi device, so I just left comments about the code itself.

junnan01-wu commented:
In latest push, I separate some patches.

junnan01-wu commented:
> okay, LGTM, but I'm not a `scmi` expert at all, so I'd like an approval from @mz-pdm. I see a conversation still open, so I'd like to wait it resolved before merging this.

Well, that's fine

mz-pdm commented on vhost-device-scmi/src/scmi.rs:
```diff
@@ -1343,6 +1483,29 @@ mod tests {
+        let notify_enabled_flag = u32::from(notify_enabled);
+        let parameters = vec![
+            MessageValue::Unsigned(sensor),
+            MessageValue::Unsigned(u32::from(notify_enabled)),
```
notify_enabled_flag can be used here.

junnan01-wu replied:
Oh yes

mz-pdm reviewed (commented):
From the SCMI point of view, the PR looks to me. I reviewed it again and added some comments, mostly about wording in comments. Once these are fixed, I can approve the PR.
I think I cannot resolve conversations for some reason, this must be done by somebody else if needed.

junnan01-wu commented on vhost-device-scmi/src/devices/common.rs:
```diff
@@ -440,19 +463,94 @@ pub trait SensorT: Send {
+        match message_id {
+            SENSOR_UPDATE => {
+                // Once there is data coming, read it first to prohibit data remained in backend.
+                // Otherwise, it will get EVENT:IN always, then frontend will receive too much interrupt.
```
Your description is better, I will modify it.

junnan01-wu commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -161,6 +165,82 @@ const UNIT_MAPPING: &[UnitMapping] = &[
+}
+
+/// Representation of an IIO channel axis's scan type.
+/// It is readed from sysfs "scan_element/<channel>_type"
```
Well

junnan01-wu commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -161,6 +165,82 @@ const UNIT_MAPPING: &[UnitMapping] = &[
+    /// The channel scan type follows the rule
+    /// If repeat > 1, "%s:%c%d/%dX%d>>%u\n"
+    /// Else, "%s:%c%d/%d>>%u\n".
+    /// For more details, lookup kernel "drivers/iio/industrialio-buffer.c"
```
well

junnan01-wu commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -175,6 +255,35 @@ struct Axis {
     /// sufficiently accurate SCMI value that is represented by an integer (not
     /// a float) + decadic exponent.
     custom_exponent: i8,
+    /// Channel scan type, when sensor supports notification, it is necessary.
```
Well

junnan01-wu commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -175,6 +255,35 @@ struct Axis {
     custom_exponent: i8,
+    /// Channel scan type, when sensor supports notification, it is necessary.
+    /// The data from /dev/iio:deviceX will be formatted according to this.
+    /// The ChanScanType is phared from "scan_elements/<channel>_type"
```
Yes, it's a typo

junnan01-wu commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -287,6 +405,103 @@ impl SensorT for IIOSensor {
+        // For example, if the content of in_XXX_type is le:s16/16>>0, each value is a little endian
+        // signed 16-bit integer. For a 3-axes sensor with [x, y, z, (t)] values, i.e. the 3 axes plus an
+        // optional timestamp, we need 6 or 8 bytes buffer.
+        // Currently, only support "le:s16/16>>0"
```
well

junnan01-wu commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -287,6 +405,103 @@ impl SensorT for IIOSensor {
+
+        let signed = scan_type.sign == 's';
+        let le_endian = scan_type.endianness == IioEndian::IioLe;
+        if !signed || !le_endian {
```
Good suggestion

junnan01-wu commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -287,6 +405,103 @@ impl SensorT for IIOSensor {
+            Ok(len) => {
+                if len > 0 {
+                    result.push(MessageValue::Unsigned(0)); //Agentid
+                    result.push(MessageValue::Unsigned(device_index)); //Sensorid
```
Ok

junnan01-wu commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -287,6 +405,103 @@ impl SensorT for IIOSensor {
+
+                    // If sensor_attributes_low Bit[9] Timestamp supported.
+                    // sensor can provide timestamped values.
+                    // Then timestamp should be processed.
```
It is clearer than before, I will follow your suggestion

junnan01-wu commented on vhost-device-scmi/src/devices/iio.rs:
```diff
@@ -287,6 +405,103 @@ impl SensorT for IIOSensor {
+                        let sensor_value_low = (value_i64 & 0xffff_ffff) as i32;
+                        let sensor_value_high = (value_i64 >> 32) as i32;
+                        result.push(MessageValue::Signed(sensor_value_low));
+                        result.push(MessageValue::Signed(sensor_value_high));
```
Well

junnan01-wu commented on vhost-device-scmi/src/scmi.rs:
```diff
@@ -575,6 +591,26 @@ pub trait ScmiDevice: Send {
+
+    /// Get notify message from this device.
+    ///
+    /// Usually need to redefine it for different protocol.
```
well

junnan01-wu commented on vhost-device-scmi/src/scmi.rs:
```diff
@@ -629,9 +669,39 @@ impl DeviceMap {
 
+pub type DeviceIdentify = (ProtocolId, usize);
+
+/// EventfdMap is a structure which used to construct the relationship between device_event and DeviceIdentify
```
well

junnan01-wu commented on vhost-device-scmi/src/scmi.rs:
```diff
@@ -629,9 +669,39 @@ impl DeviceMap {
+/// EventfdMap is a structure which used to construct the relationship between device_event and DeviceIdentify
+/// Once a device supports notification, it should insert a key-value to this hashmap
+/// "device_event" is automatically assigned according to "available device event",
+/// then function handle_event can find device via this hashmap.
```
well

junnan01-wu commented on vhost-device-scmi/src/scmi.rs:
```diff
@@ -629,9 +669,39 @@ impl DeviceMap {
+struct EventfdMap {
+    // Next available device_event, it should be initialized with NOTIFY_ALLOW_START_FD
+    available_device_event: u16,
+    // Construct hashmap about device_event -> (ProtocolID, DeviceID) to find device
```
Well

junnan01-wu commented on vhost-device-scmi/src/scmi.rs:
```diff
@@ -720,8 +851,13 @@ impl ScmiHandler {
             .expect("Impossibly large number of SCMI protocols")
     }
 
+    // If a device can notify, record it's event_fd which will be assigned to it device in future.
```
Well

junnan01-wu commented on vhost-device-scmi/src/scmi.rs:
```diff
@@ -1343,6 +1483,29 @@ mod tests {
+            result,
+            handler,
+        );
+    }
```
well

junnan01-wu commented:
It seems that I have no permission to modify the file "Cargo.lock", though I rebase on current main, the CI will failed in "cargo audit". It really confused me because I do same test in my local, and it passed.

I will revert the modificaiton of "Cargo.lock" and try rebase again.

junnan01-wu commented:
Hello all, it seems that "cargo audit" will report warning in buildkite CI for necessity to update "Cargo.lock". But when I modify "Cargo.lock" and push,  github reminds me that I have no permission. Could somebody help me?

junnan01-wu commented:
@mz-pdm @stefano-garzarella Hello, I rebase this pull request and fix the CI error. It seems good now, and could you please help approval it.

junnan01-wu commented:
> Why do you need to modify Cargo.lock just because of rebase? And it shouldn't be part of "scmi: add eventqueue notification" commit, right?

Yes, it is a part of this commit. Previous, I didn't do modificaiton on "Cargo.lock“ but no audit error occurs. 

And now, if I don't do modificaiton "Cargo.lock", it will fail in "cargo audit".
Therefore I update the "Cargo.lock“ by command "cargo generate-lockfile"

But I don't know the reason why CI pass before though I have not modified it

junnan01-wu commented:
So do you means that I should not modify this file?

mz-pdm commented:
I'm not sure what's the right thing to do, let's see what the maintainers recommend.

stefano-garzarella commented:
@junnan01-wu the audit is not about your changes, but we run it on every PR changes in the CI, so sometime it's triggered also if not related. I just opened https://github.com/rust-vmm/vhost-device/pull/724 to fix it.

Please rebase on main or on that PR avoiding updating the entire `Cargo.lock`
I think I'm not allowed to rebase your branches (for some reason), otherwise I can do it later.

junnan01-wu commented:
> @junnan01-wu the audit is not about your changes, but we run it on every PR changes in the CI, so sometime it's triggered also if not related. I just opened #724 to fix it.
> 
> Please rebase on main or on that PR avoiding updating the entire `Cargo.lock` I think I'm not allowed to rebase your branches (for some reason), otherwise I can do it later.

OK, I will revert the modificaiton of "Cargo.lock" and rebase on main.

junnan01-wu commented:
@stefano-garzarella @mz-pdm Hello, I have rebased my pull request on the latest main ( after your fix [#724](https://github.com/rust-vmm/vhost-device/pull/724), audit check passed) and all checks have passed.

Could you please help approve it again.

mz-pdm commented:
Looks good to me.

stefano-garzarella reviewed (approved):
LGTM, thanks!

junnan01-wu commented:
> please can you check setting on your fork/branch to allow force pushes by maintainers?
> Otherwise you need to do it every time, since I receive this message when I try: `Couldn't update "vhost-device-scmi": Cannot force-push to this branch`

@stefano-garzarella Oh, sorry. It is because my branch protection rules on my fork/branch. I have removed it just now. Sorry for that causing you confusion.

epilys reviewed (approved):
There are some places where the code can be more idiomatic but other than that LGTM.

junnan01-wu commented:
Thanks for all reviewers kindly help, expecting for next cooperation
